### PR TITLE
Implement gemini-style file editing tools

### LINKED ISSUE #1194: Inspect gemini-cli and implement gemini-style file editing tools
Currently we have claude-style file editing tools already implemented in this repo.

We should check the gemini CLI: https://github.com/google-gemini/gemini-cli

And see how they implement file editing tools. We can implement an alternative tool set that mimics the gemini tools and see if it works better.

### PR DESCRIPTION
**Pull request by neubig**
## Summary

This PR implements gemini-style file editing tools as an alternative to the existing claude-style `file_editor` tool.

Fixes #1194

### New Tools Implemented

Based on the analysis of [gemini-cli](https://github.com/google-gemini/gemini-cli), I implemented four separate tools instead of one command-based tool:

1. **read_file** - Read file content with pagination support
   - Supports `offset` and `limit` parameters for reading large files in chunks
   - Automatically numbers lines for better context
   - Truncates large files with clear indicators

2. **write_file** - Full file overwrite operations
   - Creates new files or overwrites existing ones
   - Automatically creates parent directories
   - Shows diff visualization between old and new content

3. **edit** - Find and replace with validation
   - Performs exact string matching and replacement
   - Optional `expected_replacements` parameter to validate the number of changes
   - Creates new files when `old_string` is empty
   - Shows diff visualization of changes

4. **list_directory** - Directory listing with metadata
   - Lists files and directories with size and modification time
   - Supports recursive listing (up to 2 levels deep)
   - Handles both relative and absolute paths

### Key Differences from Claude-Style Tools

- **Separate tools vs. commands**: Each operation is a distinct tool rather than commands within a single tool
- **More explicit**: Tool names directly indicate their purpose
- **Validation built-in**: The `edit` tool includes `expected_replacements` for validation
- **Pagination**: The `read_file` tool supports offset/limit for handling large files

### Code Organization

```
openhands-tools/openhands/tools/gemini_file_editor/
├── __init__.py           # Module exports
├── read_file.py          # Read file tool definition
├── write_file.py         # Write file tool definition
├── edit.py               # Edit tool definition
├── list_directory.py     # List directory tool definition
└── executor.py           # All executor implementations

tests/tools/gemini_file_editor/
├── test_read_file.py     # 7 test cases
├── test_write_file.py    # 6 test cases
├── test_edit.py          # 9 test cases
└── test_list_directory.py # 7 test cases
```

### Testing

All 29 test cases pass, covering:
- ✅ Basic functionality for each tool
- ✅ Error handling (file not found, invalid paths, etc.)
- ✅ Edge cases (empty files, large files, special characters)
- ✅ Pagination and truncation
- ✅ Recursive directory listing
- ✅ Find/replace validation

### Quality Checks

- ✅ All tests passing
- ✅ Type checking with pyright
- ✅ Linting with ruff
- ✅ Code formatting with ruff format
- ✅ PEP8 style compliance

## Usage Example

```python
from openhands.tools.gemini_file_editor import (
    ReadFileTool,
    WriteFileTool,
    EditTool,
    ListDirectoryTool,
)

# Read a file
read_tool = ReadFileTool.create(conv_state)
action = ReadFileAction(file_path="example.py")
obs = await read_tool.executor(action)

# Edit a file
edit_tool = EditTool.create(conv_state)
action = EditAction(
    file_path="example.py",
    old_string="def old_function():",
    new_string="def new_function():",
    expected_replacements=1,
)
obs = await edit_tool.executor(action)
```

## Future Work

This implementation can be evaluated against the existing claude-style `file_editor` to determine which approach works better for different use cases. Both tools can coexist, allowing users to choose based on their preferences and needs.

## Checklist

- [x] Implementation complete for all four tools
- [x] Comprehensive tests added (29 test cases)
- [x] All tests passing
- [x] Pre-commit checks passing (ruff, pyright, pycodestyle)
- [x] Code follows repository conventions
- [x] Documentation in docstrings and tool descriptions

@neubig can click here to [continue refining the PR](https://app.all-hands.dev/conversations/9cffb023d3e54652affbc5a2fc75b423)

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |

**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:2acad74-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-2acad74-python \
  ghcr.io/openhands/agent-server:2acad74-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:2acad74-golang-amd64
ghcr.io/openhands/agent-server:2acad74-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:2acad74-golang-arm64
ghcr.io/openhands/agent-server:2acad74-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:2acad74-java-amd64
ghcr.io/openhands/agent-server:2acad74-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:2acad74-java-arm64
ghcr.io/openhands/agent-server:2acad74-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:2acad74-python-amd64
ghcr.io/openhands/agent-server:2acad74-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-amd64
ghcr.io/openhands/agent-server:2acad74-python-arm64
ghcr.io/openhands/agent-server:2acad74-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-arm64
ghcr.io/openhands/agent-server:2acad74-golang
ghcr.io/openhands/agent-server:2acad74-java
ghcr.io/openhands/agent-server:2acad74-python
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `2acad74-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `2acad74-python-amd64`) are also available if needed

### CODE REVIEW

**Comment by enyst** [comment]
I ran the updated GPT-5.1 apply_patch example end-to-end against this branch to "
    "exercise the new tool in a more realistic, multi-file/multi-hunk scenario.

"
    "Script: `examples/01_standalone_sdk/28_apply_patch_with_gpt5_1.py`

"
    "What the agent did:
"
    "- Created `FACTS.txt` and `NOTES.md` with the requested initial contents.
"
    "- Applied a **multi-hunk** update to `NOTES.md` in a single `*** Update File` block, "
    "changing `Initial point A` -> `Updated point A` and inserting a new bullet "
    "`- Added via multi-hunk patch.` after `Initial point B`.
"
    "- Applied a **multi-file** patch that updated **both** `FACTS.txt` and `NOTES.md` in one "
    "`*** Begin Patch` ... `*** End Patch`:
"
    "  - `FACTS.txt`: appended `Multi-file patches are supported.`
"
    "  - `NOTES.md`: appended `Summary: multi-file patch applied.`
"
    "- Exercised the **mixed add / update / delete** flow for `TEMP.txt` and `FACTS.txt`:
"
    "  - First attempted to add+delete TEMP variants in a single patch, which correctly "
    "surfaced `Delete File Error: Missing File: ...` and `Duplicate Path`-style behavior.
"
    "  - Then split this into two valid patches: one to add `TEMP.txt` and append "
    "`Cleanup step ran.` to `FACTS.txt`, and a second to delete `TEMP.txt`.

"
    "Final on-disk state after the run:
"
    "- `FACTS.txt`:
"
    "  ```
"
    "  OpenHands SDK integrates tools.
"
    "  ApplyPatch can edit multiple files.
"
    "  Multi-file patches are supported.
"
    "  Cleanup step ran.
"
    "  ```
"
    "- `NOTES.md`:
"
    "  ```
"
    "  # Notes
"
    "  - Updated point A
"
    "  - Initial point B
"
    "  - Added via multi-hunk patch.
"
    "  Summary: multi-file patch applied.
"
    "  ```

"
    "This confirms that the new apply_patch implementation handles:
"
    "- multi-hunk updates in a single file,
"
    "- multi-file updates in a single patch, and
"
    "- mixed add/update/delete operations across patches, with errors like "
    "delete-on-missing-file surfaced as explicit `DiffError`s rather than silent "
    "no-ops.
"

File: openhands-tools/openhands/tools/gemini_file_editor/__init__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+from openhands.tools.gemini_file_editor.list_directory import ListDirectoryTool
+from openhands.tools.gemini_file_editor.read_file import ReadFileTool
+from openhands.tools.gemini_file_editor.write_file import WriteFileTool
+
+
+__all__ = [
+    "ReadFileTool",
+    "WriteFileTool",
+    "EditTool",
+    "ListDirectoryTool",
```
Comment (enyst) [inline openhands-tools/openhands/tools/gemini_file_editor/__init__.py]: Nit: Since they are separate tools, could we perhaps structure their code with each tool defined in its own directory, like the other tools are?

File: openhands-tools/openhands/tools/gemini_file_editor/list_directory.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                    entry_type = "📁" if entry.is_directory else "📄"
+                    size_str = (
+                        "-" if entry.is_directory else self._format_size(entry.size)
+                    )
+                    modified_str = entry.modified_time.strftime("%Y-%m-%d %H:%M")
+                    table.add_row(entry_type, entry.name, size_str, modified_str)
+
+                from io import StringIO
+
+                from rich.console import Console
```
Comment (enyst) [inline openhands-tools/openhands/tools/gemini_file_editor/list_directory.py]: This is a bit surprising, I don't know, is it necessary to use rich.console?

**Comment by neubig** [comment]
I'd really like to get this working, will try to make some time.

**Comment by neubig** [comment]
@openhands reflect enyst's comments.

Once that is done, I'd like you to run the integration tests with these new tools with the gemini-3-pro language model. I will provide you an API key.

**Comment by neubig** [comment]
This needs eval.

**Comment by neubig** [comment]
@openhands download and understand the OpenHands/benchmarks repo. Think carefully about how you can run this PR with the swe-bench benchmark, paying particular attention to what you need to do to make sure that the tools introduced in this benchmark are used instead of the default editing tools. I'll provide you with the API keys you need to perform the evaluation. Run on 50 examples first.

**Comment by neubig** [comment]
## SWE-bench Evaluation Results with Gemini-Style File Editing Tools

I ran the SWE-bench evaluation on 50 instances from `princeton-nlp/SWE-bench_Verified` using the gemini-style file editing tools introduced in this PR.

### Configuration
- **Model**: `gemini-3-pro-preview` (via litellm_proxy)
- **Tools**: Gemini-style (`read_file`, `write_file`, `edit`, `list_directory`)
- **Max iterations**: 100
- **SDK commit**: `c3ade03`

### Results Summary

| Metric | Count |
|--------|-------|
| **Total instances submitted** | 50 |
| **Instances completed** | 45 |
| **Instances resolved** ✅ | 30 |
| **Instances unresolved** ❌ | 15 |
| **Empty patches** | 4 |
| **Errors** | 1 |

### Resolution Rate: **66.7%** (30/45 completed instances)

### Resolved Instances (30)
- `astropy__astropy-12907`, `astropy__astropy-13236`, `astropy__astropy-14309`
- `django__django-11095`, `django__django-11133`, `django__django-11999`, `django__django-12050`, `django__django-12155`, `django__django-12262`, `django__django-13158`, `django__django-13279`, `django__django-13670`, `django__django-14434`, `django__django-14999`, `django__django-15278`
- `pytest-dev__pytest-5262`, `pytest-dev__pytest-5631`, `pytest-dev__pytest-5809`
- `scikit-learn__scikit-learn-12973`, `scikit-learn__scikit-learn-13439`, `scikit-learn__scikit-learn-14983`, `scikit-learn__scikit-learn-25232`, `scikit-learn__scikit-learn-25973`
- `sphinx-doc__sphinx-10466`, `sphinx-doc__sphinx-7757`, `sphinx-doc__sphinx-8551`, `sphinx-doc__sphinx-8621`
- `sympy__sympy-13877`, `sympy__sympy-15599`, `sympy__sympy-23534`

### Notes
- 4 instances had empty patches due to Docker image build failures during the image preparation step (transient infrastructure issues)
- 1 instance (`matplotlib__matplotlib-25775`) had an error during evaluation (missing Docker image)
- The gemini-style tools were successfully used by the agent throughout the evaluation

### Output Location
Full results are available at:
`outputs/gemini-tools-eval/princeton-nlp__SWE-bench_Verified-test/litellm_proxy/gemini-3-pro-preview_sdk_c3ade03_maxiter_100_N_initial/`

**Comment by enyst** [comment]
> Changes Made to Enable Gemini Tools Evaluation

IMHO this is cool, I really like what it did! 

@neubig For comparison, here is the relevant situation for GPT-5.x:
- https://github.com/OpenHands/software-agent-sdk/pull/1166
- https://github.com/OpenHands/software-agent-sdk/pull/1462

So we already have `apply_patch` tools, the GPT equivalent of this PR, merged in main. But we don't have a preset for them. I picked up the preset idea from this PR, and I had the agent implement it for GPT-5 (PR 1462):
- define and register the preset
- it's not default, even for GPTs
- the user, such as `benchmarks` client repo, or CLI, can choose the set of tools to use.

---

I see OH just posted a result here, I don't know the baseline for these 50 instances for Gemini though, so I'm not sure if it's low or high.

I'd like to suggest that if it did _work correctly_ (if logs show they ran normally, and the agent is not hallucinating) IMHO we could merge this. Because merging this PR means we will offer _everyone else_ the Gemini tools, they can decide to enable them for the agent, to see how they work, to see if they think of some improvement or run into issues. Their feedback will come back to us and enable us to move forward.

I mean, they're not on by default, even for Gemini, and I think maybe that could be okay until we see performance improvements, WDYT?

**Comment by neubig** [comment]
Thanks @enyst, I'm OK with this! I'm also going to try to run the baseline as well though.

File: examples/01_standalone_sdk/30_gemini_file_tools.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import os
+
+from openhands.sdk import LLM, Agent, Conversation, Tool
+from openhands.tools.gemini import GEMINI_FILE_TOOLS
+from openhands.tools.terminal import TerminalTool
+
+
+llm = LLM(
+    model=os.getenv("LLM_MODEL", "anthropic/claude-sonnet-4-5-20250929"),
```
Comment (enyst) [inline examples/01_standalone_sdk/30_gemini_file_tools.py]:
```suggestion
    model=os.getenv("LLM_MODEL", "gemini/gemini-3-pro-preview"),
```

**Comment by enyst** [comment]
@OpenHands Help me with reviewing this PR. It adds an example to examples/ : I need you to run this example script in your environment and see if it works, and if it worked with gemini tools indeed, not with default tools.

First, add another message from the user to the script, to tell the agent to delete that file it created.

Then try to run it with the gemini key in your env (and no base url); if that doesn't work for some reason, tweak it to use the litellm proxy with the key you have for that (and base url https://llm-proxy.eval.all-hands.dev); you can find out the correct model name prefix on the proxy. (maybe gemini/, maybe litellm_proxy/ , maybe no prefix just the model name?)

Make sure telemetry is enabled and you save logs (see llm.py); then verify the logs. Show me snippets from the logs that show what tool calls exactly did the agent make during the run. Post results.

**Comment by enyst** [comment]
HUMAN: how did it loop?

OpenHands-GPT-5:

It got stuck repeatedly calling the same tool with the same arguments, without progressing to the write/delete steps.

**What happened**
- The model issued multiple consecutive ListDirectoryAction calls with dir_path="." and recursive=False.
- Each call returned the same directory listing observation.
- The model then asked for list_directory again instead of moving on to write_file or terminal.
- After several such action→observation cycles, the stuck detector triggered and the run stopped.

**Evidence from the run**
- Console trace showed repeated “Agent Action: ListDirectoryAction … dir_path: '.' … recursive: False” three times in a row, followed by:
  - WARNING openhands.sdk.conversation.stuck_detector: Action, Observation loop detected
  - WARNING openhands.sdk.conversation.impl.local_conversation: Stuck pattern detected.
- Telemetry logs for the direct Gemini run (no base URL) all recorded the same tool call:
  - gemini__gemini-3-pro-preview-1766284619.592-91aa.json → "list_directory"
  - gemini__gemini-3-pro-preview-1766284621.939-4cfc.json → "list_directory"
  - gemini__gemini-3-pro-preview-1766284624.238-efa3.json → "list_directory"
  - gemini__gemini-3-pro-preview-1766284626.609-1f46.json → "list_directory"

In short, it looped by repeatedly selecting the list_directory tool with identical inputs, leading the stuck detector to halt the run.

**Comment by enyst** [comment]
HUMAN: 😅 

OK, I debugged this for a while, and it's very weird. Three times in a row, it got stuck, doing nothing. But then it just worked. Nothing in the logs, nothing happened, it just returned the same tool call at first (getting stuck), and then it didn't.

I think it has nothing to do with the PR. The new tools are active and executed normally.

Notes for later:
- only AIStudio Gemini got stuck; Vertex / proxy works as expected
- when it got stuck, the logs showed empty `thinking_blocks` every tool call; when it succeeded, the logs show some empty `thinking_blocks`, but some tool calls don't have them at all (?)
- Vertex logs show some non-empty `thinking_blocks`, even if it's just the tool call repeated, the LLM is using those.
